src/CommunityToolkit.Maui/Views/GravatarImageSource.shared.cs

```diff
@@ -0,0 +1,261 @@
+			}
+			else
+			{
+				double? objectWidthRequest = (double?)(element.GetType().GetProperty("WidthRequest")?.GetValue(element, null));
+				if (objectWidthRequest is not null && (double)objectWidthRequest >= 0)
+				{
+					width = (double)objectWidthRequest;
+				}
+			}
+
+			var sizeFromParent = (int)Math.Max(width, height);
+			if (GravatarSize != sizeFromParent)
+			{
+				GravatarSize = sizeFromParent;
+				HandleNewUriRequested(Email, Image);
+			}
+		}
+		catch
+		{
+			GravatarSize = defaultSize;
+		}
+	}
+
+	/// <summary>On property changed.</summary>
+	/// <param name="propertyName">Property name.</param>
+	protected override void OnPropertyChanged(string propertyName)
+	{
+		if (propertyName is not null && (propertyName == CacheValidityProperty.PropertyName || propertyName == CachingEnabledProperty.PropertyName))
+		{
+			OnUriChanged();
+		}
+
+		base.OnPropertyChanged(propertyName);
+	}
+
+	static string DefaultGravatarName(DefaultImage defaultGravatar)
+			=> defaultGravatar switch
+			{
+				DefaultImage.FileNotFound => "404",
+				DefaultImage.MysteryPerson => "mp",
+				_ => $"{defaultGravatar}".ToLower(),
+			};
+
+	static string GetMd5Hash(ReadOnlySpan<char> str)
+	{
+		if (str.Length is 0)
+		{
+			return string.Empty;
+		}
+
+		using var md5 = MD5.Create();
+		Span<byte> hash = md5.ComputeHash(Encoding.UTF8.GetBytes(str.ToArray()));
+		return BitConverter.ToString(hash.ToArray(), 0, hash.Length).Replace("-", string.Empty, StringComparison.OrdinalIgnoreCase).ToLowerInvariant();
+	}
+
+	static void OnDefaultImagePropertyChanged(BindableObject bindable, object oldValue, object newValue)
+	{
+		GravatarImageSource gravatarImageSource = (GravatarImageSource)bindable;
+		gravatarImageSource.HandleNewUriRequested(gravatarImageSource.Email, (DefaultImage)newValue);
+	}
+
+	static void OnEmailPropertyChanged(BindableObject bindable, object oldValue, object newValue)
+	{
+		GravatarImageSource gravatarImageSource = (GravatarImageSource)bindable;
+		gravatarImageSource.HandleNewUriRequested((string?)newValue, gravatarImageSource.Image);
+	}
+
+	async Task<Stream> DownloadStreamAsync(Uri? uri, CancellationToken cancellationToken)
+	{
+		try
+		{
+			return await StreamWrapper.GetStreamAsync(uri, cancellationToken, singletonHttpClient).ConfigureAwait(false);
+		}
+		catch (Exception ex)
+		{
+			Debug.WriteLine($"Error getting stream for {Uri}: {ex}");
+			return System.IO.Stream.Null;
+		}
+	}
+
+	void HandleNewUriRequested(string? email, DefaultImage image)
+	{
+		Uri = string.IsNullOrWhiteSpace(email)
+			? new Uri($"{defaultGravatarImageAddress}?s={GravatarSize}")
+			: new Uri($"{defaultGravatarImageAddress}{GetMd5Hash(email)}?s={GravatarSize}&d={DefaultGravatarName(image)}");
+
+		OnUriChanged();
+	}
+
+	void OnUriChanged()
+	{
+		if (tokenSource is not null)
+		{
+			tokenSource.Cancel();
+			tokenSource.Dispose();
+		}
+
+		tokenSource = new CancellationTokenSource();
+		Task.Delay(737, tokenSource.Token).ContinueWith(task =>
```

Comment:
Do we need to delay here?

Reply:
I'm going to do some testing around here, as the default way (copied from .NET MAUI) feels dirty and inefficient.  This other was recommended by Pedro and is used for similar in the `UserStoppedTypingBehavior`, and I need to review what I have a bit more.

Reply:
@bijington I have done some tests around this.

The original code, as taken from .NET MAUI.  When initially called, only 1 call to download the stream is made.  However, when updating multiple properties, this results in multiple calls to the download stream of which all but the last is cancelled:

```
Error getting stream for https://www.gravatar.com/avatar/04d3f95947702213e23730a0e8fac6c3?s=128&d=monsterid: System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at System.Net.Http.HttpClient.HandleFailure(Exception e, Boolean telemetryStarted, HttpResponseMessage response, CancellationTokenSource cts, CancellationToken cancellationToken, CancellationTokenSource pendingRequestsCts)
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
   at Microsoft.Maui.Controls.StreamWrapper.GetStreamAsync(Uri uri, CancellationToken cancellationToken, HttpClient client)
   at CommunityToolkit.Maui.Views.GravatarImageSource.DownloadStreamAsync(Uri uri, CancellationToken cancellationToken) in C:\Dev\MAUI\CommunityToolkitMaui\src\CommunityToolkit.Maui\Views\GravatarImageSource.shared.cs:line 165
```

Implementing the task with a delay and continue with, when initially called, only 1 call to download the stream is made.  When the same multiple properties are updated, it results in only 1 call to the download stream and no cancellations.

If however this is done without an appropriate delay, this does not catch that multiple properties have changed, and results in multiple calls to the download stream and then cancelled.

For example:
```csharp
	[RelayCommand]
	void Button_Clicked()
	{
		Email = "bleepbloop";
		DefaultGravatarSelected = DefaultImage.MonsterId;
		CacheValidityInDays = 2;
		EnableCache = false;
	}
```

This results in:
```
OnUriChanged 14:43:28.204722
OnUriChanged 14:43:28.222350
OnUriChanged 14:43:28.230345
OnUriChanged 14:43:28.232433
```

There is a measurable delay between the first property changed and the others.  Each property changes results in a download stream request, and if we build in a delay that is short enough to not cause a problem, but long enough to catch MOST updates, then this results in the best scenario for the end user.  In that we are not attempting multiple download streams, only to then cancel them.

Given the above, and the average speed of an end users device, I think a delay of ~500ms would cover most scenarios? 

Thoughts?

Reply:
Thank you for investigating this. The delay makes some sense now as it's providing debounce logic, sorry I didn't realise that was the case. I guess if this delay amount came from .NET MAUI we can probably stick with it

Reply:
The delay amount is pure my number. The code is from the stopped trying behaviour in MCT by Pictos, where the default is 1000ms.
However, looking at the average delay between property changes, when done linear in code puts it on average between 300 and 800ms, when all 4 properties (that cause a new image to be requested) are changed.

Reply:
@GeorgeLeithead can you please move this magic number to a constant? This will give some context for future contributors (;

Reply:
@pictos It IS a magic number! https://youtu.be/BUZ6b4M1Oi4?t=36 I will move it to be a constant!